firebase-firestore/src/test/java/com/google/firebase/firestore/local/IndexFreeQueryEngineTest.java

```diff
@@ -0,0 +1,251 @@
+              Query query, SnapshotVersion sinceReadTime) {
+            assertEquals(
+                "Observed query execution mode did not match expectation",
+                expectIndexFreeExecution,
+                !SnapshotVersion.NONE.equals(sinceReadTime));
+            return super.getDocumentsMatchingQuery(query, sinceReadTime);
+          }
+        };
+    queryEngine.setLocalDocumentsView(localDocuments);
+  }
+
+  /** Add a document to local cache and the remote key mapping. */
+  private void addDocumentToRemoteResult(Document doc) {
+    persistence.runTransaction(
+        "addDocumentToRemoteResult",
+        () -> {
+          queryCache.addMatchingKeys(
+              DocumentKey.emptyKeySet().insert(doc.getKey()), TEST_TARGET_ID);
+          remoteDocumentCache.add(doc, doc.getVersion());
+        });
+  }
+
+  /** Add a document to local cache but not the remote key mapping. */
+  private void addDocumentToLocalResult(Document doc) {
+    remoteDocumentCache.add(doc, doc.getVersion());
+  }
+
+  private DocumentSet runQuery(Query query, QueryData queryData, boolean expectIndexFree) {
+    expectIndexFreeExecution = expectIndexFree;
+    ImmutableSortedMap<DocumentKey, Document> docs =
+        queryEngine.getDocumentsMatchingQuery(
+            query, queryData, queryCache.getMatchingKeysForTargetId(TEST_TARGET_ID));
+    View view =
+        new View(query, new ImmutableSortedSet<>(Collections.emptyList(), DocumentKey::compareTo));
+    View.DocumentChanges viewDocChanges = view.computeDocChanges(docs);
+    return view.applyChanges(viewDocChanges).getSnapshot().getDocuments();
+  }
+
+  @Test
+  public void usesTargetMappingForInitialView() {
+    Query query = query("coll").filter(filter("matches", "==", true));
+    QueryData queryData = queryData(query, /* hasLimboFreeSnapshot= */ true);
+
+    addDocumentToRemoteResult(MATCHING_DOC_A);
+    addDocumentToRemoteResult(MATCHING_DOC_B);
+
+    DocumentSet docs = runQuery(query, queryData, /* expectIndexFree= */ true);
+    assertEquals(docSet(query.comparator(), MATCHING_DOC_A, MATCHING_DOC_B), docs);
+  }
+
+  @Test
+  public void filtersNonMatchingInitialResults() {
+    Query query = query("coll").filter(filter("matches", "==", true));
+    QueryData queryData = queryData(query, /* hasLimboFreeSnapshot= */ true);
+
+    addDocumentToRemoteResult(MATCHING_DOC_A);
+    addDocumentToRemoteResult(MATCHING_DOC_B);
+
+    addDocumentToLocalResult(PENDING_NON_MATCHING_DOC_A);
+
+    DocumentSet docs = runQuery(query, queryData, /* expectIndexFree= */ true);
+    assertEquals(docSet(query.comparator(), MATCHING_DOC_B), docs);
+  }
+
+  @Test
+  public void includesChangesSinceInitialResults() {
+    Query query = query("coll").filter(filter("matches", "==", true));
+    QueryData originalQueryData = queryData(query, /* hasLimboFreeSnapshot= */ true);
+
+    addDocumentToRemoteResult(MATCHING_DOC_A);
+    addDocumentToRemoteResult(NON_MATCHING_DOC_B);
```

Comment:
Similar to my earlier question: I don't understand how this happens. Is this shorthand for the query result initially matching B, but then an update changed B to not match?

Also, all versions of document B are at version 2 and changing `UPDATED_MATCHING_DOC_B` doesn't change that. I guess this happens to work because `queryData` forces the initial query to version 1, but that means that we never find B except because it's been updated. 

Is this the premise of the test? If so it'd be worth spelling this out in comments.

If so, note that "remote" and "local" don't really capture the difference between these operations that well since both cases are updating the remote document cache. One case is setting which documents match the query and the other is performing an out-of-band update I'd suggest renaming these `initiallyMatched` (or `matchesQuery` or `addToQueryResult`) vs  `otherQueryUpdates` (or `updateSeparately`, `updateByOtherQuery`, `updateOutOfBand`, or something similar).

Reply:
For this test case, I can just use two matching documents in the beginning and overwrite one to be non-matching. There is a similar test case where that does quite achieve the same test coverage, I added a comment here (`doesNotUseInitialResultsForLimitQueryWithDocumentRemoval()`).

Furthermore, I split the addDocumentToRemoteResult() into two calls. We now have `addDocument()` and `persistQueryMapping()` which should hopefully make it much clearer what is going on.